src/Sylius/Component/Core/Promotion/Action/ItemDiscountAction.php

```diff
@@ -0,0 +1,124 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Component\Core\Promotion\Action;
+
+use Sylius\Component\Core\Model\AdjustmentInterface;
+use Sylius\Component\Core\Model\OrderInterface;
+use Sylius\Component\Core\Model\OrderItemInterface;
+use Sylius\Component\Core\Model\OrderItemUnitInterface;
+use Sylius\Component\Originator\Originator\OriginatorInterface;
+use Sylius\Component\Promotion\Action\PromotionActionInterface;
+use Sylius\Component\Promotion\Model\PromotionInterface;
+use Sylius\Component\Promotion\Model\PromotionSubjectInterface;
+use Sylius\Component\Resource\Exception\UnexpectedTypeException;
+use Sylius\Component\Resource\Factory\FactoryInterface;
+
+/**
+ * @author Mateusz Zalewski <mateusz.zalewski@lakion.com>
+ */
+abstract class ItemDiscountAction implements PromotionActionInterface
+{
+    /**
+     * @var FactoryInterface
+     */
+    protected $adjustmentFactory;
+
+    /**
+     * @var OriginatorInterface
+     */
+    protected $originator;
+
+    /**
+     * @param FactoryInterface $adjustmentFactory
+     * @param OriginatorInterface $originator
+     */
+    public function __construct(FactoryInterface $adjustmentFactory, OriginatorInterface $originator)
+    {
+        $this->adjustmentFactory = $adjustmentFactory;
+        $this->originator = $originator;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public abstract function execute(PromotionSubjectInterface $subject, array $configuration, PromotionInterface $promotion);
```

Comment:
I would suggest to keep abstract functions together. And maybe just after constructor?